### ISSUE TITLE
Suppress notebook output and show output in connected console?

### ISSUE THREAD

**Issue by h4gen** [jupyterlab/jupyterlab#4700]
Hey everybody!
I am looking for the RStudio/Matlab/Spyder experience. Therefore I want to use the notebook primarily to prototype and structure the code in logical units. I guess, I am not the only one who thinks notebooks get pretty confusing with all the outputs over time.
Since jupyterlab allows to connect a console to a notebook, it would be awesome to suppress the notebook output if a console is connected which displays the output. 
Right now the output ist just doubled.
Is this already possible?
If not, I would like to help implementing it if someone shows me where to start!

Cheers!

**Comment by BoPeng** [comment]
This is implemented in [this PR](https://github.com/jupyterlab/jupyterlab/pull/4330). One of the questions from the developers was how useful this feature is so perhaps you should make your voice heard there. :smile:

**Comment by h4gen** [comment]
Thanks for the information @BoPeng . I will make a comment there :) I know a lot of people who think of this feature as useful!

**Comment by h4gen** [comment]
@BoPeng I see your PR differs a bit from my Question. As far as I can see you propose a line by line execution? Would this also include to execute whole cells without selecting the whole code in the cell first?

**Comment by aldanor** [comment]
I had a somewhat related question: https://github.com/jupyterlab/jupyterlab/issues/4671

Basically, would be nice to have something like

```python
%%preview
do_some_stuff_that_logs_a_ton_of_lines()
# and see the output magically appear in a separate output view
```

**Comment by BoPeng** [comment]
@h4gen Yes, selecting the entire cell would be needed to execute the cell in console using my PR. However, we already have a number of actions/shortcuts for executing cell content and it is difficult to justify two shortcuts for executing in console. My understanding is that “console” is designed to be a scratch area (result not saved in notebook) so running code line by line for debugging is more important than running entire cell there.

@h4gen and @aldanor :  Another possibility for these kind of scenario is #4550 where a separate UI is used to display transient information. For the ‘execute in console’ case, a magic could be used to direct the output to this UI as transient result (not in notebook). For the `%%preview` case, the kernel could send status information to the UI continuously during the execution of the long execution or send long debug information to this UI. The [SoS kernel](https://vatlab.github.io/sos-docs/) for classic Jupyter can already [`%preview`](https://vatlab.github.io/sos-docs/doc/documentation/Notebook_Interface.html#Preview-of-results--14), [`%clear`](https://vatlab.github.io/sos-docs/doc/documentation/SoS_Magics.html#magic-clear), [`%render`](https://vatlab.github.io/sos-docs/doc/documentation/SoS_Magics.html#magic-render) or [`%capture`](https://vatlab.github.io/sos-docs/doc/documentation/SoS_Magics.html#magic-capture) results after the completion of a cell and it would be natural to have a magic to redirect results to the side panel after the implementation of #4550.  Again, I received no response on #4550 which made me wonder if this is useful to others.... If you are interested, you could [open a ticket on SoS Notebook](https://github.com/vatlab/sos-notebook/issues) and propose a magic to redirect cell output to side panel.

**Comment by saulshanabrook** [comment]
If there was a keyboard shortcut to execute the cell and make the output collapsed, would this work for you? It seems like this could be a non-invasive way to achieve the desired effect.

**Comment by h4gen** [comment]
@saulshanabrook This would be awesome!

**Comment by saulshanabrook** [comment]
OK if anyone wants to implement that, it would be a good first issue. I am happy to help walk you through anything if you get stuck.

**Comment by jasongrout** [comment]
In order for the shortcut to happen, we'd need a new notebook execution command that would execute and collapse. Here is the code to run a notebook cell: https://github.com/jupyterlab/jupyterlab/blob/bf808f8fa61ac9aabcde19ac00f6b61a9e6295a0/packages/notebook/src/actions.tsx#L1443

**Comment by BoPeng** [comment]
@h4gen How exactly do you want this command to behave? The run-in-console action executes content in the console so the execution count of the notebook does not change. I think you are looking more like a run-cell-with-output-in-console action so that the notebook cell is executed with input shown in both notebook and console (with the same execution count) and output shown only in console.

**Comment by h4gen** [comment]
@BoPeng Yeah that sums it up pretty much! I don't know if it would be necessary to have the input in the console again (So much redundancy) but it would be already a great feature this way!

**Comment by BoPeng** [comment]
This is useful when I want to "execute" the cell without leaving any trace. I can even imagine some zealous support for this feature since it prompts some R-studio style of working when the main notebook is always "clean" and can be git-able and diff-able, and the output is in a separate place (console) especially if it is easy to save the console output as a separate notebook file (with the downside of not having any markdown).

There many ways to implement this 

1. **A flag or an action**. It could be implemented as a notebook option "Send output to console", that affects all executions including `run-all`. It is a bit difficult to use when you would like to execute only one cell this way, but it helps you a lot if you want to use the notebook as the "script" and console as "output". In my case, many of my cells generate long output which makes the notebook difficult to navigate and such a mode can be very useful. Actually, I found myself use `run-in-console` a lot just to reduce output in notebook.

2. **Whether or not change execution count of notebook cell**. If you have the mentality of executing the notebook with output sent to console, the execution count should be changed. If you are thinking of executing in the console, the execution count should not be changed, and there is a corner case if you want the original execution count to be removed in this case. A "diff-able" notebook should not have any execution count though.

3. **Whether or not remove existing output**. Existing output should be kept in a "run-in-console" style of execution when input/output happens in the console and the notebook cell is untouched. Existing output should be removed if the execution happens (logically) in notebook with execution count updated.

4. **Whether or not add input to console**. Again, if you are "sending output to console", namely if you are treating the console as a log book then you do not want the repetition (and the input execution count of the notebook has to change). However, it can be easier and clearer if you include the input cell in console.

In the end, I found myself leaning towards a flag that affects all cell execution actions, in which mode the cell is executed in console with input and execution count, and it removes execution count and existing output (if any) of the notebook. This allows me to keep the notebook clean and small, making it diff-able and easier to navigate but it makes it a three step process if you only want to apply the mode to one cell... :smile:
